File: solutions/c/dnd-character/1/dnd_character.c
#include "dnd_character.h"
#include <math.h>
#include <stdlib.h>
#include <time.h>

#define DICE 4
#define SIDES 6
#define MODIFIER 10
#define HITPOINTS 10

int ability(void) {
    int sum = 0;
    int min = SIDES + 1;
    
    for (int i = 0; i < DICE; i++) {
        int roll = (rand() % SIDES) + 1;
        if (roll < min) min = roll;
        sum += roll;
    }
    
    return sum - min;
}

int modifier(int score) {
    return (int)floor((score - MODIFIER) / 2.0);
}

dnd_character_t make_dnd_character(void) {
    srand(time(NULL));
    dnd_character_t character = {0};

    character.strength     = ability();
    character.dexterity    = ability();
    character.constitution = ability();
    character.intelligence = ability();
    character.wisdom       = ability();
    character.charisma     = ability();
    character.hitpoints    = HITPOINTS + modifier(character.constitution);

    return character;
}

File: solutions/elixir/dnd-character/2/lib/dnd_character.ex
defmodule DndCharacter do

  @type t :: %__MODULE__{
    strength: integer,
    dexterity: integer,
    constitution: integer,
    intelligence: integer,
    wisdom: integer,
    charisma: integer,
    hitpoints: integer
  }

  defstruct ~w[strength dexterity constitution intelligence wisdom charisma hitpoints]a

  @spec modifier(integer) :: integer
  def modifier(score) when score >= 10, do: div(score - 10, 2)
  def modifier(score), do: Integer.floor_div(score - 10, 2)

  @spec ability() :: integer
  def ability() do
    1..4
    |> Enum.map(fn _ -> Enum.random(1..6) end)
    |> Enum.sort()
    |> Enum.drop(1)
    |> Enum.sum()
  end

  @spec character() :: t()
  def character() do
    constitution = ability()

    %__MODULE__{
      strength: ability(),
      dexterity: ability(),
      constitution: constitution,
      intelligence: ability(),
      wisdom: ability(),
      charisma: ability(),
      hitpoints: 10 + modifier(constitution)
    }
  end
  
end
